app.js

```diff
@@ -0,0 +1,187 @@
+//f the user is successfully found, the user object is passed as the second argument to the done function, indicating successful deserialization. we pass 'null' as an argument- when there is no error
+passport.deserializeUser(async function (id, done) {
+  try {
+    const user = await User.findById(id);
+    done(null, user);
+  } catch (err) {
+    done(err);
+  }
+});
+
+app.use(
+  session({
+    secret: process.env.SESSION_SECRET,
+    resave: false,
+    saveUninitialized: true,
+    store: store,
+  })
+);
+
+// sets up Passport to work with Express and prepares it for authentication handling.
+app.use(passport.initialize());
+//enables Passport to restore authentication state from the session, allowing authenticated users to stay logged in across multiple requests.
+app.use(passport.session());
+//parse URL-encoded data from incoming requests to handle form submissions that send data in the application/x-www-form-urlencoded format.
+//extended: false option indicates that the URL-encoded data should be parsed using the built-in querystring library of Node.js
+app.use(express.urlencoded({ extended: false }));
+//in other words express.urlencoded() middleware enables our application to handle URL-encoded form data.
+
+// sets up a variable currentUser in the res.locals object, making it available in our views for convenient access to the currently authenticated user information.
+app.use(function (req, res, next) {
+  res.locals.currentUser = req.user;
+  next();
+});
+
+const authMiddleware = (req, res, next) => {
+  if (!req.user) {
+    if (!req.session.messages) {
+      req.session.messages = [];
+    }
+    req.session.messages.push("You can't access that page before logon.");
+    res.redirect('/');
+  } else {
+    next();
+  }
+};
+
+//show error messages to the user . The messages are put in an array, req.session.messages.
+app.get('/', (req, res) => {
+  let messages = [];
+  if (req.session.messages) {
+    messages = req.session.messages;
+    req.session.messages = [];
+  }
+  res.render('index', { messages });
+});
+
+app.get('/sign-up', (req, res) => res.render('sign-up-form'));
+
+//delete all the session information at logoff time
+app.get('/log-out', (req, res) => {
+  req.session.destroy(function (err) {
+    res.redirect('/');
+  });
+});
+
+// HIDE PASSWORD
+app.post('/sign-up', async (req, res, next) => {
+  try {
+      const { username }= req.body;
+      const usernameAlreadyExists = await User.findOne({username})
+      if(usernameAlreadyExists){
+        req.session.messages.push("This username is already taken. Please choose another one");
+        res.redirect('/');
+      }
+    const hashedPassword = await bcrypt.hash(req.body.password, 10);
+    await User.create({
+      username: req.body.username,
+      password: hashedPassword,
+    });
+    res.redirect('/');
+  } catch (err) {
+    return next(err);
```

Comment:
If we want to use the `mongoose-unique-validator` library; it basically cleans up that "E11000 duplicate key error" so that it's a little easier to find the error and a nice message for it. 

If we use that library, then here in the `catch` block we can do something a little different:

```javascript
  } catch (err) {
    console.log("HANDLING SIGN UP ERROR: ", err?.errors)

    // If the error that we're catching is due to someone trying to sign up with a duplicate username
    // then we will set a message and redirect them
    // otherwise we pass the error along to the next middleware using `next(...)`
    if (err.name === 'ValidationError') {
      if (!req.session.messages) {
        req.session.messages = [];
      }
      req.session.messages.push(err?.errors?.username?.message);
      req.session.messages.push(err?.errors?.password?.message);
      return res.redirect('/sign-up');
    }
    return next(err);
  }
```

The error thrown by the library is named `ValidationError` and apart from the usual fields a Javascript error has (name and message) it also has some additional fields as shown in the [npm docs](https://www.npmjs.com/package/mongoose-unique-validator). Here is what my console log line shows:

```
HANDLING SIGN UP ERROR:  {
  username: ValidatorError: Error, expected `username` to be unique. Value: `akos`
      at validate (/Users/aasante/dev/passport-demo/node_modules/mongoose/lib/schematype.js:1347:13)
      at /Users/aasante/dev/passport-demo/node_modules/mongoose/lib/schematype.js:1323:24
      at processTicksAndRejections (node:internal/process/task_queues:96:5) {
    properties: {
      message: 'Error, expected `username` to be unique. Value: `akos`',
      type: 'unique',
      validator: [Function (anonymous)],
      path: 'username',
      value: 'akos'
    },
    kind: 'unique',
    path: 'username',
    value: 'akos',
    reason: undefined,
    [Symbol(mongoose:validatorError)]: true
  }
}
```

Reply:
If you're not familiar with the `?.` syntax; it's called optional chaining: [https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining). It just let's you dig into an inner field in a nested object without having to manually do an if check for the presence of each field.

Reply:
@akosasante I've played around with both ways to set up a message, with and without the mongoose-unique-validator library, and it worked! Thanks again!